ZEPXXX-priority-election.md

```diff
@@ -0,0 +1,231 @@
+
+
+#### Multi-raft aware raft partition node
+
+A raft partition node knows about other partitions in the same node.
+On election timeout, the partition checks if the number of leaders in the node is below a threshold. If yes, it starts election. Otherwise it doesn't trigger election.
+Thus the number of leaders in the node is kept under the threshold.
+
+We did not chose this approach because:
+1. A partition is aware of other partitions.
+
+#### On bootstrap primary starts election immediately
+To ensure that on cluster start the primary becomes the leader, we can let the primary starts sending poll requests immediately after it is started.
+This might trigger an election even if there is an existing leader.
+It is possible to implement this behavior together with the priority election.
+
+
+#### Priority based election timeout
+Instead of randomized election timeout, the election timeout are assigned based on the priority.
+The node with highest priority has a lower timeout, the node with lowest priority has the highest timeout.
+This way the highest priority node will start the election first and thus has a higher chance of becoming the leader.
+We did not chose this approach because it is almost similar to the priority election discussed in this document. We do not see any advantage over priority election. There might be corner cases which we have to test and optimize.
+Since the priority election is already implemented and tested in other raft implementation, we decided not to investigate more time in this approach.
+
+We decided to chose priority election because:
+1. The solution can be implemented with in a raft partition.
+2. It is easy to implement.
+3. No external trigger required.
+4. In future it gives us the opportunity to make the priority configurable and thus a user can control how the leaders are assigned.
+
+# Prior art
+[prior-art]: #prior-art
+
+The priority election proposed in this document is implemented in another raft based system - sofa-jraft. https://www.sofastack.tech/blog/sofa-jraft-priority-election/
+
+# Out of scope
+[out-of-scope]: #out-of-scope
+
+* Exposing an api to manually balance leaders
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+1. Reduce election duration. Some ideas:
+    - If the current leader is the primary, the secondary can immediately starts the election if the primary dies. No need to wait until the priority is reduced.
+    - After the first election timeout, the timeouts in the next rounds can be reduced. There is no advantage in waiting for another electionTimeout.
+2. Exposing an api to trigger leader distribution.
+3. Allow users to configure priorities.
+4. Dynamically assign priorities based on other properties. For example, if a node is overloaded it may reduce the priority for some partitions so that it doesn't become the leader.
```

Comment:
Plus: On bootstrap primaries start immediately to poll?

Reply:
:+1: